torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -99,11 +102,16 @@ void DistEngine::computeDependencies(
   edge_list recvBackwardEdges;
   // Traverse the graph.
   auto& dependencies = graphTask->dependencies_;
+  // If a node is in this map, it should be replaced by its corresponding value
+  // in the graph, i.e. all edges pointing to the key should point to the value.
+  std::unordered_map<Node*, std::shared_ptr<DistAccumulateGrad>>
+      accumulate_grad_replacements;
   while (!queue.empty()) {
     auto fn = queue.front();
     queue.pop();
 
-    for (const auto& edge : fn->next_edges()) {
+    for (int index = 0; index < fn->num_outputs(); ++index) {
+      const auto& edge = fn->next_edge(index);
       if (auto nextFn = edge.function.get()) {
```

Comment:
Why do we need a raw pointer here? Is it just to avoid the overhead of reference counting? IIUC, `edge.function` is a shared_ptr.

Reply:
Legacy code. The reasons seem
1. The shared_ptr may be nullptr. Therefore, an `if` is here.
2. All containers in this function, like `queue` and `dependencies `, use raw pointers. It's more convenient to have a raw pointer.